### PR TITLE
Local Signup

### PR DESCRIPTION
**Pull request by spankie**
The signup is working. Please review and gimme your comment. Its a small commit but just want you to know I'm still working on it. (at least one pull request everyday)

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -41,8 +41,8 @@ func Init() {
 
 	// set mongo server to default if it is not set as a system variable
 	if MONGOSERVER == "" {
-		//MONGOSERVER = "mongodb://spankie:506dad@ds163738.mlab.com:63738/btshopng"
-		MONGOSERVER = "127.0.0.1:27017"
+		MONGOSERVER = "mongodb://spankie:506dad@ds163738.mlab.com:63738/btshopng"
```
Comment (tonyalaribe) [inline config/config.go]: Install mongo db in your development environment so you can use the local ip, or you add the mongoserver to your environment path. 

I prefer you have a local mongo instance for debugging purposes. What if you want to quickly check if the data is actually entering the db or something?  You want to start logging into mongo lab?

File: web/utils.go [hunk 2]
```diff
@@ -34,6 +37,21 @@ func Userget(r *http.Request) (models.User, error) {
 	return user, nil
 }
 
+func encrypt(word string) (string, error) {
```
Comment (tonyalaribe) [inline web/utils.go]: This function is unneccesary. When storing password, you simply run  
`
hashedWord, err := bcrypt.GenerateFromPassword([]byte(word), bcrypt.DefaultCost)
`

and store the hashedPassword in the database as a binary. Not as a string. 

Then during login, you use
`
isMatch := bcrypt.CompareHashAndPassword(hashedWord, []byte(word))
`
The hasedWord would be the ninary password that was stored in the database.while word is the password text inserted by the user.

Bottom line is, there's no point for the two to be in the same function. Its just resource wastage. Dont forget encryption and decryption are not big O free operations.

File: web/auth.go [hunk 3]
```diff
@@ -58,6 +71,83 @@ func SignupPageHandler(w http.ResponseWriter, r *http.Request) {
...
+	r.ParseForm()
+
+	fullName := r.FormValue("name")
+	email := r.FormValue("email")
+	passwd := r.FormValue("passwd")
+
+	// Check if the Post data not empty and validate them.
+	if fullName != "" && email != "" && passwd != "" {
+		// encrypt password.
+		password, _ := encrypt(passwd)
```
Comment (tonyalaribe) [inline web/auth.go]: just run
`
password, err := bcrypt.GenerateFromPassword([]byte(passwd), bcrypt.DefaultCost)
`

here

File: web/auth.go [hunk 4]
```diff
@@ -58,6 +71,83 @@ func SignupPageHandler(w http.ResponseWriter, r *http.Request) {
...
+	data.FBAuthURL = FBOauthConf.AuthCodeURL("state", oauth2.AccessTypeOffline)
+
+	r.ParseForm()
+
+	fullName := r.FormValue("name")
+	email := r.FormValue("email")
+	passwd := r.FormValue("passwd")
+
+	// Check if the Post data not empty and validate them.
+	if fullName != "" && email != "" && passwd != "" {
```
Comment (tonyalaribe) [inline web/auth.go]: revert this line. The bulk of the function should not be in an if clause. Instead run 

`
if fullName==""||email==""||passwd==""{
   data.SignupError = "Please Fill out all required Fields"
 	tmp := GetTemplates().Lookup("signin_signup.html")
 	tmp.Execute(w, data)
        return
}
`
then remove the current content of the function into the default scope.

Learn the habit of keeping the larger code area in the root scope.  Else you'd quickly have unmaintainable code

**Review by tonyalaribe — CHANGES_REQUESTED** [review]
Install mongo db in your development environment so you can use the local ip, or you add the mongoserver to your environment path.

File: config/config.go (hunk 1 above)
Comment (spankie) [reply config/config.go]: I've done a lot to install mongo db on my system. even tried using another distro(Kali linux). The binaries it needs to run are hard to find, I really do not have the time now for it. I've really tried alot to make it work. I would just make it a system variable...I really do not mind logging in to mlab... Thanks

File: web/utils.go (hunk 2 above)
Comment (spankie) [reply web/utils.go]: Got you.

btw: I just realized you are critical about some theoretical computer science topics (Big O). thats cool.

File: web/auth.go (hunk 4 above)
Comment (spankie) [reply web/auth.go]: Good Idea

File: web/auth.go (hunk 3 above)
Comment (spankie) [reply web/auth.go]: Got it.

File: models/user.go [hunk 5]
```diff
@@ -76,3 +76,48 @@ func (user User) Get(c *config.Conf) (User, error) {
 	}
 	return user, nil
 }
+
+// CheckUser checks if a user exists in the database
+func (user User) CheckUser(c *config.Conf) (User, error) {
```
Comment (tonyalaribe) [inline models/user.go]: CheckUser and GetUser do the exact same thing. I think this should be removed, and getuser used in its place.

File: models/user.go [hunk 6]
```diff
@@ -75,3 +76,48 @@ func (user User) Get(c *config.Conf) (User, error) {
...
+	// log.Println("USER: ", user, "result:", result)
+	// if result.Name == "" {
+	// 	return result, errors.New("Username or Password is incorrect(no user)")
+	// }
+
+	return result, nil
+}
+
+// Insert : inserts user data to the DB
+func (user User) Insert(c *config.Conf) error {
```
Comment (tonyalaribe) [inline models/user.go]: Does almost same things as upsert. Problem is, using insert like that will cause duplicates. What if I log in with with fb, and my email is in the db. if I create an account again and you store it with insert, or without an index to prevent duplicates, itd be problematic. It would even be better to extend the upsert function to instead add any missing data that the new request has. eg if I create an email and password account, instead of not adding the data, it would update the database field, so I can login with both email and fb, and still be the same user. For regular users, generate a UUID as the document ID. for fb, I'm using the user ID facebook returns.

File: web/auth.go [hunk 7]
```diff
@@ -58,6 +72,202 @@ func SignupPageHandler(w http.ResponseWriter, r *http.Request) {
...
+	// }
+
+	// log.Println("USER: ", nuser)
+	// data.LoginMessage = "You are logged in"
+	// tmp := GetTemplates().Lookup("signin_signup.html")
+	// tmp.Execute(w, data)
+}
+
+// SignupHandler handles the signup process for in app signup
+func SignupHandler(w http.ResponseWriter, r *http.Request) {
```
Comment (tonyalaribe) [inline web/auth.go:155]: This handler is doing way more than it should be doing.  It should only process the form values, and redirect back /signup if theres any issue. not re rendering /signup content at its route. If you're passing error message, then pass it via the url e.g redirect to /signup?error=bla+bla+bla&errortype=1 somthing like that. thats a quick example. but you can read the url query params and render the error from the SignupPageHandler. So this should just be for form processing and redirection. Not for rerendering the same page multiple times

Comment (tonyalaribe) [reply web/auth.go:155]: I even thing we might rename the route from /signup to something else, since both login and signup forms are on the same page.

File: web/auth.go [hunk 8]
```diff
@@ -58,6 +72,202 @@ func SignupPageHandler(w http.ResponseWriter, r *http.Request) {
...
+	// I feel SignupPageHandler() and this function should share the same struct for better memory conservation
+	data := struct {
```
Comment (tonyalaribe) [inline web/auth.go]: no need for this

File: web/auth.go [hunk 9]
```diff
@@ -58,6 +72,202 @@ func SignupPageHandler(w http.ResponseWriter, r *http.Request) {
...
+		LoginMessage  string
+		SignupError   string
+	}{}
+
+	data.LoginMessage = ""
+	data.SignupError = ""
+
+	// This is set here so that when there are any errors from the signup process,
+	// the link will be passed to the template alongside the errors.
+	data.FBAuthURL = FBOauthConf.AuthCodeURL("state", oauth2.AccessTypeOffline)
```
Comment (tonyalaribe) [inline web/auth.go]: no need

File: web/auth.go [hunk 10]
```diff
@@ -58,6 +72,202 @@ func SignupPageHandler(w http.ResponseWriter, r *http.Request) {
...
+	fullName := r.FormValue("name")
+	email := r.FormValue("email")
+	passwd := r.FormValue("passwd")
+
+	// Check if the Post data not empty and validate them.
+	if fullName == "" && email == "" && passwd == "" {
+
+		data.SignupError = "Please Fill out all required Fields"
+		tmp := GetTemplates().Lookup("signin_signup.html")
+		tmp.Execute(w, data)
```
Comment (tonyalaribe) [inline web/auth.go]: redirect instead of rendering the page here

File: web/auth.go [hunk 11]
```diff
@@ -58,6 +72,202 @@ func SignupPageHandler(w http.ResponseWriter, r *http.Request) {
...
+		FormattedDateCreated: now.String(),
+		Password:             password,
+	}
+
+	// CHECK IF THE EMAIL HAS ALREADY BEEN USED BEFORE.
+	_, err := user.CheckUser(config.GetConf())
+	if err == nil {
+		data.SignupError = "Email address has already been used."
+		tmp := GetTemplates().Lookup("signin_signup.html")
+		tmp.Execute(w, data)
```
Comment (tonyalaribe) [inline web/auth.go]: redirect

File: web/auth.go [hunk 12]
```diff
@@ -58,6 +72,202 @@ func SignupPageHandler(w http.ResponseWriter, r *http.Request) {
...
+	// 	data.SignupError = "Could not Sign you up rigt now. Try Again"
+	// 	tmp := GetTemplates().Lookup("signin_signup.html")
+	// 	tmp.Execute(w, data)
+	// 	return
+	// }
+	err = user.Insert(config.GetConf())
+	if err != nil {
+		data.SignupError = "Could not sign you up. Try again."
+		tmp := GetTemplates().Lookup("signin_signup.html")
+		tmp.Execute(w, data)
```
Comment (tonyalaribe) [inline web/auth.go]: redirect. By the way, its also ok to create a separate error route for pages errors that need their own page, if necessary. Then redirect there. You could also redirect to a success page, depending on the case

File: web/auth.go [hunk 13]
```diff
@@ -58,6 +72,202 @@ func SignupPageHandler(w http.ResponseWriter, r *http.Request) {
...
+	}
+	// note: Userget(r) is passing a string to User.Password instead of []byte
+
+	// SHOULD VERIFY EMAIL ADDRESS SENT, HERE.
+
+	// If all goes well, generate a token for the cookie
+	loginResp, err := GenerateJWT(user)
+	if err != nil {
+		log.Println(err)
+		http.Redirect(w, r, "/signup", 301)
```
Comment (tonyalaribe) [inline web/auth.go:212]: if redirecting due to an error. Add an error param, so you can handle the eerror appropriately, whether by add error message or anything.

File: web/auth.go [hunk 14]
```diff
@@ -58,6 +72,202 @@ func SignupPageHandler(w http.ResponseWriter, r *http.Request) {
...
+		return
+	}
+	// Create cookie
+	expire := time.Now().AddDate(0, 0, 1)
```
Comment (tonyalaribe) [inline web/auth.go:217]: The same token should be used. Its basically an encoded json object containing the users data (the struct)

Comment (tonyalaribe) [reply web/auth.go:217]: you should read about JWT (Json web tokens. And know what makes them better than regular sessions)

**Review by tonyalaribe — CHANGES_REQUESTED** [review]
Nice improvements. Try to understand the reasons behind most decisions.  If you dont, you can ask me. Let me know when you've made the changes.  Theres a lot to be done. lets not spent days on just authentication

File: models/user.go (hunk 5 above)
Comment (spankie) [reply models/user.go]: okay...

File: models/user.go (hunk 6 above)
Comment (spankie) [reply models/user.go]: The upsert() function replaces the document in the collection even if their values are different, But what i didn't try is adding uuid. Maybe it will add new values if the uuid is included. I'll try it out and make it work.

File: web/auth.go (hunk 7 above)
Comment (spankie) [reply web/auth.go:155]: Okay, I thought of renaming it too, but i decided to focus on making the login work first.

I'll use the get parameters.

I don't know if you've noticed, but if {{.LoginMessage}} is not set, the form does not display at all. That's the reason why i intialize them to empty string at the beginning of the handlers. Using the GET parameters to pass those messages might just eliminate the issue.

File: web/auth.go (hunk 8 above)
Comment (spankie) [reply web/auth.go]: Yep.

File: web/auth.go (hunk 9 above)
Comment (spankie) [reply web/auth.go]: Yep.

File: web/auth.go (hunk 10 above)
Comment (spankie) [reply web/auth.go]: Yep.

File: web/auth.go (hunk 11 above)
Comment (spankie) [reply web/auth.go]: Yep.

File: web/auth.go (hunk 12 above)
Comment (spankie) [reply web/auth.go]: Will do.

File: web/auth.go (hunk 13 above)
Comment (spankie) [reply web/auth.go:212]: Okay...

File: web/auth.go (hunk 14 above)
Comment (spankie) [reply web/auth.go:217]: Okay, Will do that.

File: models/user.go (hunk 6 above)
Comment (tonyalaribe) [reply models/user.go]: like i said in the message, what could be done is to rewrite the upsert function to insert new content if it doesnt exist, or update the existing content with the new data. ie if user exists, and email doesnt exist, add email. basically update the fields that have changed. It might be possible using $set in upsert. but you could research it, or even use multiple database calls if necessary. to request the data and compare changes. If you just add data without checks, whether with new UUID or not, you could have two accounts pointing to one user. one from email signup and the other from email

File: web/auth.go (hunk 7 above)
Comment (tonyalaribe) [reply web/auth.go:155]: you have to initialize the values before they can be used. But let only one handler be in charge of rendering the login/signup page.

Comment (spankie) [reply web/auth.go:155]: On it.

File: models/user.go (hunk 6 above)
Comment (spankie) [reply models/user.go]: Okay.

**Review by tonyalaribe — APPROVED** [review]

**Comment by tonyalaribe** [comment]
Is the entire signup and login working?  What task are you working on next?

**Comment by spankie** [comment]
I'm modifying the Upsert() function as you suggested.

What do you think we should work on next?

**Comment by tonyalaribe** [comment]
Work on the new barter form submition. Dont border about file upload for
now. I'd take care of that. But let the form be entering the db, and then
display the latest barters on their profile and latest barters on the
homepage. Then work on the search and search result.  I'd be tidying up the
messaging area. Then I'd add image  uploading when you're done.

Anthony Alaribe

**Comment by spankie** [comment]
Okay, Cool.
